cpp/ql/src/semmle/code/cpp/ir/implementation/IRType.qll

```diff
@@ -0,0 +1,257 @@
+  final override string toString() { result = "int" + byteSize.toString() }
+
+  final override Language::LanguageType getCanonicalLanguageType() {
+    result = Language::getCanonicalSignedIntegerType(byteSize)
+  }
+}
+
+/**
+ * An unsigned two's-complement integer. Also used to represent enums whose underlying type is an
+ * unsigned integer, as well as character types whose representation is unsigned.
+ */
+class IRUnsignedIntegerType extends IRNumericType, TIRUnsignedIntegerType {
+  final override string toString() { result = "uint" + byteSize.toString() }
+
+  final override Language::LanguageType getCanonicalLanguageType() {
+    result = Language::getCanonicalUnsignedIntegerType(byteSize)
+  }
+}
+
+/**
+ * A floating-point type.
+ */
+class IRFloatingPointType extends IRNumericType, TIRFloatingPointType {
+  final override string toString() { result = "float" + byteSize.toString() }
+
+  final override Language::LanguageType getCanonicalLanguageType() {
+    result = Language::getCanonicalFloatingPointType(byteSize)
+  }
+}
+
+/**
+ * An address type, representing the memory address of data. Used to represent pointers, references,
+ * and lvalues, include those that are garbage collected.
+ *
+ * The address of a function is represented by the separate `IRFunctionAddressType`.
+ */
+class IRAddressType extends IRSizedType, TIRAddressType {
+  final override string toString() { result = "addr" + byteSize.toString() }
+
+  final override Language::LanguageType getCanonicalLanguageType() {
+    result = Language::getCanonicalAddressType(byteSize)
+  }
+}
+
+/**
+ * An address type, representing the memory address of code. Used to represent function pointers,
+ * function references, and the target of a direct function call.
+ */
+class IRFunctionAddressType extends IRSizedType, TIRFunctionAddressType {
+  final override string toString() { result = "func" + byteSize.toString() }
+
+  final override Language::LanguageType getCanonicalLanguageType() {
+    result = Language::getCanonicalFunctionAddressType(byteSize)
+  }
+}
+
+/**
+ * A type with known size that does not fit any of the other kinds of type. Used to represent
+ * classes, structs, unions, fixed-size arrays, pointers-to-member, and more.
+ */
+class IRBlobType extends IRSizedType, TIRBlobType {
+  Language::BlobTypeTag tag;
+
+  IRBlobType() {
+    this = TIRBlobType(tag, byteSize)
+  }
+
+  final override string toString() {
+    result = "blob" + byteSize.toString() + "{" + tag.toString() + "}"
+  }
+
+  final override string getIdentityString() {
+    result = "blob" + byteSize.toString() + "{" + Language::getBlobTagIdentityString(tag) + "}"
+  }
+
+  final override Language::LanguageType getCanonicalLanguageType() {
+    result = Language::getCanonicalBlobType(tag, byteSize)
+  }
+
+  /**
+   * Gets the "tag" that differentiates this type from other incompatible blob types that have the
+   * same size.
+   */
+  final Language::BlobTypeTag getTag() { result = tag }
```

Comment:
Is this tag used for anything now? I'm not sure it's very useful. As long as the "pointer type" is just "address" rather than "address of this IRType", we can't even get to the tag if it's hidden behind one layer of indirection.

Reply:
Right now, it's used to ensure that the `IRType` for `struct A { int i; }` and the `IRType` for `struct B { float f; }` are distinct. `IRType` equality is used to determine whether the overlap between a def and a use is `MustExactlyOverlap` or `MustTotallyOverlap`. The idea is that whatever state you might be tracking for a given definition is probably incorrect at a use that reinterprets the bits as a different type.

Reply:
I see. So right now we'll say that in C code, `struct A { int i; }` and `struct B { int i; }` are incompatible even though they aren't, and in C++ we'll say that `int *` is compatible with `double &` even though they aren't. That's probably okay for the first iteration.

Reply:
Right. I'm not sure we'll ever get to the point where either of those examples makes a difference to anything we care about.